[EPG Static Port] Make VLAN parameter optional when the state is absent (DCNE-300)

noppanut15 opened the pull request:
### Description

- Updated the code to make the `vlan` parameter optional when the `state` is `absent`.
- This change ensures that the `vlan` parameter is not required when the state is set to `absent`, preventing potential errors or unnecessary requirements.
- Added appropriate checks to handle the absence of the `vlan` parameter gracefully.

### New or Affected Module(s):

<!--- Please list the new or affected module(s). --->

* `mso_schema_site_anp_epg_staticport`

### MSO version and MSO Platform

* Tested with NDO 4.1


### APIC version and APIC Platform for Site Level Resources

* Tested with APIC 5.2

### Collection versions

* cisco.mso 2.9.0

### References
* Resolves #607

akinross commented on plugins/modules/mso_schema_site_anp_epg_staticport.py:
```diff
@@ -441,9 +441,14 @@ def main():
 
     if not static_ports and state in ["present", "absent"]:
-        key_list = ["pod", "leaf", "path", "vlan"]
+        if state == "absent":
+          # The state absent requires at least the pod, leaf and path to be provided
+          key_list = ["pod", "leaf", "path"]
+        else:
+          # The state present requires all the key_list to be provided
+          key_list = ["pod", "leaf", "path", "vlan"]
```
could you add a test case that tests that we can delete without vlan in https://github.com/CiscoDevNet/ansible-mso/blob/master/tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml?

noppanut15 replied:
@akinross Sure, I will add a test case and will fix the indentation as well.

noppanut15 commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -512,6 +513,85 @@
     that:
     - nm_remove_again_stat2e2 is not changed
 
+# ADD STATIC PORT WITHOUT VLAN
+- name: Add static port 3 (dpc) to EPG6 of AP2 without VLAN (error)
+  cisco.mso.mso_schema_site_anp_epg_staticport:
+    <<: *mso_info
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    site: '{{ mso_site | default("ansible_test") }}'
+    template: Template 1
+    anp: AP2
+    epg: EPG6
+    pod: pod-2
+    leaf: 102
+    path: eth1/3
+    deployment_immediacy: lazy
+    mode: regular
+    type: dpc
+    primary_micro_segment_vlan: 199
+    state: present
+  ignore_errors: true
+  register: nm_add_stat3e6
+
+- name: Verify nm_add_stat3e6
+  ansible.builtin.assert:
+    that:
+    - nm_add_stat3e6 is failed
+    - nm_add_stat3e6.msg == "state is present or absent but all of the following are missing{{":"}} vlan."
```
This test case is to ensure that the change won't affect the `present` state. (`vlan` is still needed)

noppanut15 commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -512,6 +513,85 @@
+    - nm_add_stat3e6 is failed
+    - nm_add_stat3e6.msg == "state is present or absent but all of the following are missing{{":"}} vlan."
+
+# ADD STATIC PORT WITH VLAN
+- name: Add static port 3 (dpc) to EPG6 of AP2 with VLAN (success)
+  cisco.mso.mso_schema_site_anp_epg_staticport:
+    <<: *mso_info
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    site: '{{ mso_site | default("ansible_test") }}'
+    template: Template 1
+    anp: AP2
+    epg: EPG6
+    pod: pod-2
+    leaf: 102
+    path: eth1/3
+    vlan: 100
+    deployment_immediacy: lazy
+    mode: regular
+    type: dpc
+    primary_micro_segment_vlan: 199
+    state: present
+  register: nm_add_stat3e6
+
```
This test case is to ensure that the change won't affect the `present` state. (`vlan` is still needed). And this port will be used for deletion in the next test case.

noppanut15 commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -512,6 +513,85 @@
+    - nm_add_stat3e6 is failed
+    - nm_add_stat3e6.msg == "state is present or absent but all of the following are missing{{":"}} vlan."
+
+# ADD STATIC PORT WITH VLAN
+- name: Add static port 3 (dpc) to EPG6 of AP2 with VLAN (success)
+  cisco.mso.mso_schema_site_anp_epg_staticport:
+    <<: *mso_info
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    site: '{{ mso_site | default("ansible_test") }}'
+    template: Template 1
+    anp: AP2
+    epg: EPG6
+    pod: pod-2
+    leaf: 102
+    path: eth1/3
+    vlan: 100
+    deployment_immediacy: lazy
+    mode: regular
+    type: dpc
+    primary_micro_segment_vlan: 199
+    state: present
+  register: nm_add_stat3e6
+
+- name: Verify nm_add_stat3e6
+  ansible.builtin.assert:
+    that:
+    - nm_add_stat3e6 is changed
+    - nm_add_stat3e6.previous == {}
+    - nm_add_stat3e6.current.deploymentImmediacy == 'lazy'
+    - nm_add_stat3e6.current.portEncapVlan == 100
+    - nm_add_stat3e6.current.microSegVlan == 199
+    - nm_add_stat3e6.current.path == 'topology/pod-2/paths-102/pathep-[eth1/3]'
+    - nm_add_stat3e6.current.mode == 'regular'
+    - nm_add_stat3e6.current.type == 'dpc'
```
This test case is to ensure that the change won't affect the `present` state. (`vlan` is still needed). And this port will be used for deletion in the next test case.

noppanut15 commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -512,6 +513,85 @@
+    - nm_add_stat3e6.current.mode == 'regular'
+    - nm_add_stat3e6.current.type == 'dpc'
+  
+# REMOVE STATIC PORT WITHOUT VLAN
+- name: Remove static port 3 (dpc) from EPG6 of AP2 without VLAN (success)
+  cisco.mso.mso_schema_site_anp_epg_staticport:
+    <<: *mso_info
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    site: '{{ mso_site | default("ansible_test") }}'
+    template: Template 1
+    anp: AP2
+    epg: EPG6
+    pod: pod-2
+    leaf: 102
+    path: eth1/3
+    state: absent
+  register: nm_add_stat3e6
+
+- name: Verify nm_add_stat3e6
+  ansible.builtin.assert:
+    that:
+    - nm_add_stat3e6 is changed
+    - nm_add_stat3e6.current == {}
```
Check if we can remove the static port without VLAN

noppanut15 commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -512,6 +513,85 @@
+    - nm_add_stat3e6.current.mode == 'regular'
+    - nm_add_stat3e6.current.type == 'dpc'
+  
+# REMOVE STATIC PORT WITHOUT VLAN
+- name: Remove static port 3 (dpc) from EPG6 of AP2 without VLAN (success)
+  cisco.mso.mso_schema_site_anp_epg_staticport:
+    <<: *mso_info
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    site: '{{ mso_site | default("ansible_test") }}'
+    template: Template 1
+    anp: AP2
+    epg: EPG6
+    pod: pod-2
+    leaf: 102
+    path: eth1/3
+    state: absent
+  register: nm_add_stat3e6
+
+- name: Verify nm_add_stat3e6
+  ansible.builtin.assert:
+    that:
+    - nm_add_stat3e6 is changed
+    - nm_add_stat3e6.current == {}
+    
```
Check if we can remove the static port without VLAN

noppanut15 commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -1171,6 +1251,137 @@
     - nm_remove_static_ports_2.current.9.path == "topology/pod-1/paths-101/pathep-[eth2/9]"
     - nm_remove_static_ports_2.current.9.portEncapVlan == 1209
 
+# ADD STATIC PORT WITHOUT VLAN (BULK)
+- name: Add static port 3 & 4 (dpc) to epg_bulk of anp_bulk without VLAN (error)
+  cisco.mso.mso_schema_site_anp_epg_staticport:
+    <<: *mso_info
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    site: '{{ mso_site | default("ansible_test") }}'
+    template: template_bulk
+    anp: anp_bulk
+    epg: epg_bulk
+    static_ports:
+      - pod: pod-2
+        leaf: 102
+        path: eth1/3
+        deployment_immediacy: lazy
+        mode: regular
+        type: dpc
+        primary_micro_segment_vlan: 199
+      - pod: pod-2
+        leaf: 102
+        path: eth1/4
+        deployment_immediacy: lazy
+        mode: regular
+        type: dpc
+        primary_micro_segment_vlan: 199
+    state: present
+  ignore_errors: true
+  register: nm_add_stat3e6
+
+- name: Verify nm_add_stat3e6
+  ansible.builtin.assert:
+    that:
+    - nm_add_stat3e6 is failed
+    - nm_add_stat3e6.msg == "state is present but all of the following are missing{{":"}} vlan."
```
[TEST IN BULK] This test case is to ensure that the change won't affect the `present` state. (`vlan` is still needed)

noppanut15 commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -1171,6 +1251,137 @@
+        leaf: 102
+        path: eth1/4
+        vlan: 100
+        deployment_immediacy: lazy
+        mode: regular
+        type: dpc
+        primary_micro_segment_vlan: 199
+    state: present
+  register: nm_add_bulk_stat3e6
+
+- name: Verify nm_add_bulk_stat3e6
+  ansible.builtin.assert:
+    that:
+    - nm_add_bulk_stat3e6 is changed
+    - nm_add_bulk_stat3e6.previous | length == 10
+    - nm_add_bulk_stat3e6.previous.0.path == "topology/pod-1/paths-101/pathep-[eth1/1]"
+    - nm_add_bulk_stat3e6.previous.0.portEncapVlan == 1101
+    - nm_add_bulk_stat3e6.previous.2.path == "topology/pod-1/paths-101/pathep-[eth1/5]"
+    - nm_add_bulk_stat3e6.previous.2.portEncapVlan == 1105
+    - nm_add_bulk_stat3e6.previous.4.path == "topology/pod-1/paths-101/pathep-[eth1/9]"
+    - nm_add_bulk_stat3e6.previous.4.portEncapVlan == 1109
+    - nm_add_bulk_stat3e6.previous.5.path == "topology/pod-1/paths-101/pathep-[eth2/1]"
+    - nm_add_bulk_stat3e6.previous.5.portEncapVlan == 1201
+    - nm_add_bulk_stat3e6.previous.7.path == "topology/pod-1/paths-101/pathep-[eth2/5]"
+    - nm_add_bulk_stat3e6.previous.7.portEncapVlan == 1205
+    - nm_add_bulk_stat3e6.previous.9.path == "topology/pod-1/paths-101/pathep-[eth2/9]"
+    - nm_add_bulk_stat3e6.previous.9.portEncapVlan == 1209
+    - nm_add_bulk_stat3e6.current | length == 12
+    - nm_add_bulk_stat3e6.current.10.deploymentImmediacy == 'lazy'
+    - nm_add_bulk_stat3e6.current.10.portEncapVlan == 100
+    - nm_add_bulk_stat3e6.current.10.microSegVlan == 199
+    - nm_add_bulk_stat3e6.current.10.path == 'topology/pod-2/paths-102/pathep-[eth1/3]'
+    - nm_add_bulk_stat3e6.current.10.mode == 'regular'
+    - nm_add_bulk_stat3e6.current.10.type == 'dpc'
+    - nm_add_bulk_stat3e6.current.11.deploymentImmediacy == 'lazy'
+    - nm_add_bulk_stat3e6.current.11.portEncapVlan == 100
+    - nm_add_bulk_stat3e6.current.11.microSegVlan == 199
+    - nm_add_bulk_stat3e6.current.11.path == 'topology/pod-2/paths-102/pathep-[eth1/4]'
+    - nm_add_bulk_stat3e6.current.11.mode == 'regular'
+    - nm_add_bulk_stat3e6.current.11.type == 'dpc'
```
[TEST IN BULK] This test case is to ensure that the change won't affect the `present` state. (`vlan` is still needed). And this static port will be used for deletion in the next test case.

noppanut15 commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -1171,6 +1251,137 @@
+    - nm_add_bulk_stat3e6.current.11.mode == 'regular'
+    - nm_add_bulk_stat3e6.current.11.type == 'dpc'
+  
+# REMOVE STATIC PORT WITHOUT VLAN (BULK)
+- name: Remove static port 3 & 4 (dpc) from epg_bulk of anp_bulk without VLAN (success)
+  cisco.mso.mso_schema_site_anp_epg_staticport:
+    <<: *mso_info
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    site: '{{ mso_site | default("ansible_test") }}'
+    template: template_bulk
+    anp: anp_bulk
+    epg: epg_bulk
+    static_ports:
+      - pod: pod-2
+        leaf: 102
+        path: eth1/3
+      - pod: pod-2
+        leaf: 102
+        path: eth1/4
+    state: absent
+  register: nm_rm_bulk_stat3e6
+
+- name: Verify nm_rm_bulk_stat3e6
+  ansible.builtin.assert:
+    that:
+    - nm_rm_bulk_stat3e6 is changed
+    - nm_rm_bulk_stat3e6.current | length == 10
+    - nm_add_bulk_stat3e6.current.0.path == "topology/pod-1/paths-101/pathep-[eth1/1]"
+    - nm_add_bulk_stat3e6.current.0.portEncapVlan == 1101
+    - nm_add_bulk_stat3e6.current.2.path == "topology/pod-1/paths-101/pathep-[eth1/5]"
+    - nm_add_bulk_stat3e6.current.2.portEncapVlan == 1105
+    - nm_add_bulk_stat3e6.current.4.path == "topology/pod-1/paths-101/pathep-[eth1/9]"
+    - nm_add_bulk_stat3e6.current.4.portEncapVlan == 1109
+    - nm_add_bulk_stat3e6.current.5.path == "topology/pod-1/paths-101/pathep-[eth2/1]"
+    - nm_add_bulk_stat3e6.current.5.portEncapVlan == 1201
+    - nm_add_bulk_stat3e6.current.7.path == "topology/pod-1/paths-101/pathep-[eth2/5]"
+    - nm_add_bulk_stat3e6.current.7.portEncapVlan == 1205
+    - nm_add_bulk_stat3e6.current.9.path == "topology/pod-1/paths-101/pathep-[eth2/9]"
+    - nm_add_bulk_stat3e6.current.9.portEncapVlan == 1209
+
```
[TEST IN BULK] Check if we can remove the 2 static ports without VLAN

noppanut15 commented:
Hi @akinross, 

I've added the integration tests for both single and bulk operations along with the comment above. Feel free to give feedback or changes. Thanks!

noppanut15 commented:
Hi @akinross, 

I've added the integration tests for both single and bulk operations along with the comments above. Feel free to give feedback or let me know if you need any changes. Thanks!

akinross commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -512,6 +513,85 @@
     that:
     - nm_remove_again_stat2e2 is not changed
 
+# ADD STATIC PORT WITHOUT VLAN
```
this test seems redundant as this behaviour is already tested

akinross commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -512,6 +513,85 @@
+    - nm_add_stat3e6.msg == "state is present or absent but all of the following are missing{{":"}} vlan."
+
+# ADD STATIC PORT WITH VLAN
+- name: Add static port 3 (dpc) to EPG6 of AP2 with VLAN (success)
```
Could you add a clear name of what this static port is used for? We assume success and only (error) on the names usually

akinross replied:
could you create a new section at the bottom ( when the schemas and required parent objects have not been removed ) that has your tests together with a header like ## WITHOUT VLAN REMOVAL TEST ##

akinross commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -512,6 +513,85 @@
+    type: dpc
+    primary_micro_segment_vlan: 199
+    state: present
+  register: nm_add_stat3e6
```
could you use more meaningful names for the registered vars

akinross commented on tests/integration/targets/mso_schema_site_anp_epg_staticport/tasks/main.yml:
```diff
@@ -512,6 +513,85 @@
+    state: present
+  register: nm_add_stat3e6
+
+- name: Verify nm_add_stat3e6
```
assert for add seems redundant as we already test creation with its error and this has not changed in your contribution

noppanut15 commented:
I've updated the test and rebased the code to catch up with the upstream. 

Also, tested with my NDO.

<img width="992" alt="image" src="https://github.com/user-attachments/assets/b73191cc-fd58-421a-9371-10ce576b2a40" />

akinross reviewed (approved):
LGTM

gmicol reviewed (approved):
LGTM

shrsr reviewed (approved):
LGTM

samiib reviewed (approved):
LGTM

sajagana reviewed (approved):
LGTM!

lhercot reviewed (approved):
LGTM

lhercot commented:
Thank you for your contribution!
